frontend/src/metabase/collections/components/CollectionContent/CollectionItemsTable.tsx

```diff
@@ -0,0 +1,200 @@
+import type Database from "metabase-lib/v1/metadata/Database";
+import type {
+  Bookmark,
+  Collection,
+  CollectionId,
+  CollectionItem,
+  CollectionItemModel,
+} from "metabase-types/api";
+import { SortDirection, type SortingOptions } from "metabase-types/api/sorting";
+
+import {
+  CollectionEmptyContent,
+  CollectionTable,
+} from "./CollectionContent.styled";
+
+export type CollectionItemsTableProps = {
+  collectionId: CollectionId;
+} & Partial<{
+  bookmarks: Bookmark[];
+  clear: () => void;
+  collection: Collection;
+  createBookmark: CreateBookmark;
+  databases: Database[];
+  deleteBookmark: DeleteBookmark;
+  getIsSelected: (item: CollectionItem) => boolean;
+  handleCopy: (items: CollectionItem[]) => void;
+  handleMove: (items: CollectionItem[]) => void;
+  hasPinnedItems: boolean;
+  loadingPinnedItems: boolean;
+  models: CollectionItemModel[];
+  pageSize: number;
+  selectOnlyTheseItems: (items: CollectionItem[]) => void;
+  selected: CollectionItem[];
+  toggleItem: (item: CollectionItem) => void;
+  onClick: (item: CollectionItem) => void;
+  showActionMenu: boolean;
+}>;
+
+export const CollectionItemsTable = ({
+  collectionId,
+  collection,
+  getIsSelected,
+  selectOnlyTheseItems,
+  databases,
+  bookmarks,
+  createBookmark,
+  deleteBookmark,
+  loadingPinnedItems,
+  hasPinnedItems,
+  selected,
+  toggleItem,
+  clear,
+  handleMove,
+  handleCopy,
+  pageSize = COLLECTION_PAGE_SIZE,
+  models = ALL_MODELS,
+  onClick,
+  showActionMenu = true,
+}: CollectionItemsTableProps) => {
+  const [unpinnedItemsSorting, setUnpinnedItemsSorting] =
+    useState<SortingOptions>({
+      sort_column: "name",
+      sort_direction: SortDirection.Asc,
+    });
+
+  const [total, setTotal] = useState<number | null>(null);
+
+  const { handleNextPage, handlePreviousPage, setPage, page, resetPage } =
+    usePagination();
+
+  useEffect(() => {
+    if (collectionId) {
+      resetPage();
+      setTotal(null);
+    }
+  }, [collectionId, resetPage]);
+
+  const handleUnpinnedItemsSortingChange = useCallback(
+    (sortingOpts: SortingOptions) => {
+      setUnpinnedItemsSorting(sortingOpts);
+      setPage(0);
+    },
+    [setPage],
+  );
+
+  const showAllItems = !collection || isRootTrashCollection(collection);
+
+  const unpinnedQuery = {
+    collection: collectionId,
+    models,
+    limit: pageSize,
+    offset: pageSize * page,
+    ...(showAllItems ? {} : { pinned_state: "is_not_pinned" }),
+    ...unpinnedItemsSorting,
+  };
+
+  return (
+    <Search.ListLoader
+      query={unpinnedQuery}
+      loadingAndErrorWrapper={false}
+      keepListWhileLoading
+      wrapped
+    >
+      {({
+        list: unpinnedItems = [],
+        metadata = {},
+        loading: loadingUnpinnedItems,
+      }: {
+        list: CollectionItem[];
+        metadata: { total?: number };
+        loading: boolean;
+      }) => {
+        if (metadata.total) {
+          setTotal(metadata.total);
```

Comment:
I wonder what's the point of storing `total` in a state here. Seem we just set it and use it right away. There's one place where we set it to `null`, but to just have it reset to `metadata.total` again in this render block.

Also, it feels very bizarre setting a state inside a render function. I don't think we should ever do this.

Reply:
I should have left a comment here. The problem here is that `metadata` resets to `undefined` when loading a new query. For example, clicking `Next page` causes `metadata` to reset, and the `total` value on pagination resets to `null` causing some flickering. I put this in a state to preserve the value during loads

Reply:
You should probably use `onLoaded` instead of setting a state inside a render function

https://github.com/metabase/metabase/blob/f880f46d7035f586a8aec0ebe0e7eff5ff1e2880/frontend/src/metabase/entities/containers/EntityListLoader.jsx#L122

can you try if that works?

Reply:
oooh nice find. let me try that